src/tests.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+#[cfg(test)]
+mod test {
+    use crate::key_gen::Phase;
+    use crate::member::NodeID;
+    use crate::Member;
+    use bincode::serialize;
+    use itertools::Itertools;
+    use quic_p2p::Config;
+    use rand::{thread_rng, Rng, RngCore};
+    use std::collections::{btree_map::BTreeMap, btree_set::BTreeSet, HashMap};
+    use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    const NODE_NUM: usize = 7;
+    const THRESHOLD: usize = 5;
+    pub const IP_LOCAL_HOST: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+
+    fn setup_members_and_init_dkg<R: RngCore>(
+        num: usize,
+        non_responsives: BTreeSet<usize>,
+        rng: &mut R,
+    ) -> (Vec<NodeID>, Vec<Member>) {
+        let (ids, mut members, _addrs, map) = create_members(num, rng);
+
+        connect_and_initialize_dkg(&mut members, map, non_responsives, THRESHOLD);
+
+        (ids, members)
+    }
+
+    fn create_members<R: RngCore>(
+        num: usize,
+        mut rng: &mut R,
+    ) -> (
+        Vec<NodeID>,
+        Vec<Member>,
+        Vec<SocketAddr>,
+        HashMap<NodeID, SocketAddr>,
+    ) {
+        let mut member_vec = vec![];
+        let mut id_vec = vec![];
+        let mut socket_addr = vec![];
+        let mut map = HashMap::new();
+
+        for _ in 0..num {
+            let mut config = Config::default();
+            config.ip = Some(IP_LOCAL_HOST);
+            config.port = Some(0);
```

Comment:
could use a port deduced from index to make them different value?

Reply:
It can sometimes be overlapping with already occupied ports. Therefore setting `Some(0)` as port in the config makes Quicp2p use random ports provided by the OS avoiding such conflicts.